namer/consul/src/main/scala/io/buoyant/namer/consul/LookupCache.scala

```diff
@@ -33,37 +38,66 @@ private[consul] class LookupCache(
   private[this] val cachedCounter = service.counter("cached")
   private[this] val serviceStats = SvcAddr.Stats(service)
 
-  private[this] val cachedLookup: (String, SvcKey, Path, Path) => Activity[NameTree[Name]] =
-    untupled(Memoize[(String, SvcKey, Path, Path), Activity[NameTree[Name]]] {
-      case (dc, key, id, residual) =>
-        val addrFuture: Future[Var[Addr]] = resolveDc(dc).join(domain).map {
-          case ((dcName, domainOption)) =>
-            SvcAddr(
-              consulApi,
-              LookupCache.DefaultBackoffs,
-              dcName,
-              key,
-              domainOption,
-              consistency = consistency,
-              preferServiceAddress = preferServiceAddress,
-              weights,
-              serviceStats
-            )
-        }
+  /*
+   * use a shared mutex only on write, the ConcurrentHM impl should guarantee
+   * happens-before semantic of reads wrt to updates (as per docs)
+   */
+  private[this] val lookupStatusMu = new {}
+  private[this] val lookupStatus = new ConcurrentHashMap[Path, InstrumentedBind]()
+
+  private[consul] def status: Map[Path, InstrumentedBind] = lookupStatus.asScala.toMap
+
+  private[this] val lookupAddress: (Path, String, SvcKey) => Future[InstrumentedVar[Addr]] =
+    (raw, dc, key) => {
+      val pollState = SvcAddr.mkConsulPollState
+      val addrFuture: Future[InstrumentedVar[Addr]] = resolveDc(dc).join(domain).map {
```

Comment:
I think we want to reorder this so that the work of creating this Var[Addr] is only done if it's definitely not in cache.

Reply:
this would happen only if there's a race condition to compute the address, and never again once the cache is filled

are you suggesting to synchronize the whole method with the mutex for the cache and double check at the beginning of the method?